<a href="https://colab.research.google.com/github/vis-hal-k/Data_science_self-Practice/blob/main/Coursera_course/%23C1-IntroToTensorFlow%26ML-AI-DL/week2_assignment%20/MNIST_Fashion_callback_Lab2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ungraded Lab: Using Callbacks to Control Training

In this lab, you will use the [Callbacks API](https://keras.io/api/callbacks/) to stop training when a specified metric is met. This is a useful feature so you won't need to complete all epochs when this threshold is reached. For example, if you set 1000 epochs and your desired accuracy is already reached at epoch 200, then the training will automatically stop. Let's see how this is implemented in the next sections.


## Load and Normalize the Fashion MNIST dataset

Like the previous lab, you will use the Fashion MNIST dataset again for this exericise.
And also as mentioned before, you will normalize the pixel values to help optimize the training.

In [ ]:
import tensorflow as tf
fmnist = tf.keras.datasets.fashion_mnist
(x_train , y_train),(x_test,y_test) = fmnist.load_data()
x_train = x_train/255.0
x_test = x_test/255.0


4422102/4422102 [==============================] - 1s 0us/step


## Creating a Callback Class
You can create a callback by defining a class that inherits the [tf.keras.callbacks.Callback](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback) base class. From there, you can define available methods to set where the callback will be executed. For instance below, you will use the [on_epoch_end()](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback#on_epoch_end) method to check the loss at each training epoch.

In [ ]:
class myCallback(tf.keras.callbacks.Callback):
  def on_epoch_end(self , epoch , logs={}):
    '''
    Halts the training after reaching 60 percent accuracy
    Args:
      epoch (integer) - index of epoch (required but unused in the function definition below)
      logs (dict) - metric result from the training epoch
    '''
    # check accuracy
    if(logs.get('loss') < 0.4 ):
      print("\n Loss is lower than 0.4 so cancelling training! ")
      self.model.stop_training = True
# Instantiate class
callbacks = myCallback()


### Define And Compile the model

In [ ]:
# Define the model
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(512, activation=tf.nn.relu),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

# Compile the model
model.compile(optimizer=tf.optimizers.Adam(),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

### Train the model
Now you are ready to train the model. To set the callback, simply set the `callbacks` parameter to the `myCallback` instance you declared before. Run the cell below and observe what happens.

In [ ]:
# Train the model with a callback
model.fit(x_train,y_train, epochs = 10 , callbacks = [callbacks])

Epoch 1/10
1875/1875 [==============================] - 15s 8ms/step - loss: 0.4737 - accuracy: 0.8292
Epoch 2/10
1871/1875 [============================>.] - ETA: 0s - loss: 0.3558 - accuracy: 0.8699
 Loss is lower than 0.4 so cancelling training! 
1875/1875 [==============================] - 14s 7ms/step - loss: 0.3557 - accuracy: 0.8699


You will notice that the training does not need to complete all 10 epochs. By having a callback at each end of the epoch, it is able to check the training parameters and compare if it meets the threshold you set in the function definition. In this case, it will simply stop when the loss falls below `0.40` after the current epoch.

*Optional Challenge: Modify the code to make the training stop when the accuracy metric exceeds 60%.*

That concludes this simple exercise on callbacks!